File: src/main/java/cz/vhromada/catalog/validator/GenreValidator.java
package cz.vhromada.catalog.validator;

import cz.vhromada.catalog.entity.Genre;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.validator.AbstractMovableValidator;
import cz.vhromada.validation.result.Event;
import cz.vhromada.validation.result.Result;
import cz.vhromada.validation.result.Severity;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

/**
 * A class represents validator for genre.
 *
 * @author Vladimir Hromada
 */
@Component("genreValidator")
public class GenreValidator extends AbstractMovableValidator<Genre, cz.vhromada.catalog.domain.Genre> {

    /**
     * Creates a new instance of GenreValidator.
     *
     * @param genreService service for genres
     * @throws IllegalArgumentException if service for genres is null
     */
    @Autowired
    public GenreValidator(final MovableService<cz.vhromada.catalog.domain.Genre> genreService) {
        super("Genre", genreService);
    }

    /**
     * Validates genre deeply.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * </ul>
     *
     * @param data   validating genre
     * @param result result with validation errors
     */
    @Override
    protected void validateDataDeep(final Genre data, final Result<Void> result) {
        if (data.getName() == null) {
            result.addEvent(new Event(Severity.ERROR, "GENRE_NAME_NULL", "Name mustn't be null."));
        } else if (!StringUtils.hasText(data.getName())) {
            result.addEvent(new Event(Severity.ERROR, "GENRE_NAME_EMPTY", "Name mustn't be empty string."));
        }
    }

}

File: src/main/kotlin/cz/vhromada/catalog/validator/GenreValidator.kt
package cz.vhromada.catalog.validator

import cz.vhromada.catalog.entity.Genre
import cz.vhromada.common.result.Event
import cz.vhromada.common.result.Result
import cz.vhromada.common.result.Severity
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.validator.AbstractMovableValidator
import org.springframework.stereotype.Component

/**
 * A class represents validator for genre.
 *
 * @author Vladimir Hromada
 */
@Component("genreValidator")
class GenreValidator(genreService: MovableService<cz.vhromada.catalog.domain.Genre>) : AbstractMovableValidator<Genre, cz.vhromada.catalog.domain.Genre>("Genre", genreService) {

    /**
     * Validates genre deeply.
     * <br></br>
     * Validation errors:
     *
     *  * Name is null
     *  * Name is empty string
     *
     * @param data   validating genre
     * @param result result with validation errors
     */
    override fun validateDataDeep(data: Genre, result: Result<Unit>) {
        when {
            data.name == null -> {
                result.addEvent(Event(Severity.ERROR, "GENRE_NAME_NULL", "Name mustn't be null."))
            }
            data.name.isBlank() -> {
                result.addEvent(Event(Severity.ERROR, "GENRE_NAME_EMPTY", "Name mustn't be empty string."))
            }
        }
    }

}
